core/src/main/java/io/questdb/cairo/map/MapFactory.java

```diff
@@ -37,46 +37,98 @@ public class MapFactory {
+    public static Map createOrderedMap(
+            CairoConfiguration configuration,
+            @Transient @NotNull ColumnTypes keyTypes,
+            @Transient @Nullable ColumnTypes valueTypes
+    ) {
+        final int keyCapacity = configuration.getSqlSmallMapKeyCapacity();
+        final int pageSize = configuration.getSqlSmallMapPageSize();
+        return new OrderedMap(
+                pageSize,
+                keyTypes,
+                valueTypes,
+                keyCapacity,
+                configuration.getSqlFastMapLoadFactor(),
+                configuration.getSqlMapMaxResizes()
+        );
+    }
+
+    /**
+     * Creates an unordered Map pre-allocated to a small capacity to be used in GROUP BY queries, but not only.
+     * <p>
+     * The returned map may actually preserve insertion order, i.e. it may be ordered, depending on the types
+     * of key and value columns.
+     */
+    public static Map createUnorderedMap(
             CairoConfiguration configuration,
             @Transient @NotNull ColumnTypes keyTypes,
             @Transient @Nullable ColumnTypes valueTypes
     ) {
         final int keyCapacity = configuration.getSqlSmallMapKeyCapacity();
         final int pageSize = configuration.getSqlSmallMapPageSize();
-        final CharSequence mapType = configuration.getDefaultMapType();
-        if (Chars.equalsLowerCaseAscii(mapType, "fast")) {
-            return new FastMap(
-                    pageSize,
-                    keyTypes,
-                    valueTypes,
-                    keyCapacity,
-                    configuration.getSqlFastMapLoadFactor(),
-                    configuration.getSqlMapMaxResizes()
-            );
+
+        final int keySize = totalSize(keyTypes);
+        final int valueSize = totalSize(keyTypes);
+        if (keySize > 0) {
+            if (keySize <= Long.BYTES && Long.BYTES + valueSize <= Misc.CACHE_LINE_SIZE) {
+                return new Unordered8Map(
+                        keyTypes,
+                        valueTypes,
+                        keyCapacity,
+                        configuration.getSqlFastMapLoadFactor(),
+                        configuration.getSqlMapMaxResizes()
+                );
+            } else if (keySize <= 2 * Long.BYTES && 2 * Long.BYTES + valueSize <= Misc.CACHE_LINE_SIZE) {
+                return new Unordered16Map(
+                        keyTypes,
+                        valueTypes,
+                        keyCapacity,
+                        configuration.getSqlFastMapLoadFactor(),
+                        configuration.getSqlMapMaxResizes()
+                );
+            }
+        }
+
+        return new OrderedMap(
+                pageSize,
+                keyTypes,
+                valueTypes,
+                keyCapacity,
+                configuration.getSqlFastMapLoadFactor(),
+                configuration.getSqlMapMaxResizes()
+        );
+    }
+
+    private static int totalSize(ColumnTypes types) {
```

Comment:
nit: it would be good to add some comments as to how this method is expected to behave. It bails out when column is var size and -1 is a special value.